In [2]:
#! pip install selenium

In [16]:
#! pip install webdriver-manager


In [124]:
from selenium import webdriver
from datetime import datetime, timedelta
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.common.keys import Keys
from time import time
from time import sleep
import pandas as pd
from random import randint

import regex as re


import requests
from bs4 import BeautifulSoup 
from time import sleep
import pandas as pd
import numpy as np
import numpy.random as random
import requests
import json
from joblib import Parallel,delayed
from datetime import datetime

In [ ]:
website = """
########################################################
#          WEBSITE: https://veganuary.com/recipes/categories/vegan-seafood-recipes/         #
########################################################
"""
print(website)
start_time = datetime.now()
print('Crawl starting time : {}' .format(start_time.time()))
print()

articles_data = []

driver = webdriver.Chrome()
driver.get(
    "https://veganuary.com/recipes/categories/vegan-seafood-recipes/")
sleep(randint(8,11))


all_articles = driver.find_elements_by_xpath("//*div[@class='card card--recipe']")

for article in all_articles:
    article_link = article.find_element_by_css_selector('a').get_attribute('href')

articles_data.append(article_link)

driver.quit() 


In [ ]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""
print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

category_links = []

driver = webdriver.Chrome()

driver.get("https://veganuary.com/recipes/")

sleep(5)

all_category_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/categories/")]')


for link in all_category_links:
    category_links.append(link.get_attribute('href'))

driver.quit()

category_links = list(set(category_links))

df = pd.DataFrame(category_links, columns=["Category Links"])

csv_file = 'veganuary_recipe_category_links.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)


In [2]:
categories = pd.read_csv("veganuary_recipe_category_links.csv")
categories.rename(columns = {"Category Links": "category_links"}, inplace=True)
categories["category"] = categories["category_links"]

In [3]:
categories["category"] 

0     https://veganuary.com/recipes/categories/itali...
1     https://veganuary.com/recipes/categories/quick...
2     https://veganuary.com/recipes/categories/north...
3     https://veganuary.com/recipes/categories/japan...
4     https://veganuary.com/recipes/categories/vegan...
5     https://veganuary.com/recipes/categories/korea...
6     https://veganuary.com/recipes/categories/south...
7     https://veganuary.com/recipes/categories/ameri...
8     https://veganuary.com/recipes/categories/india...
9     https://veganuary.com/recipes/categories/frenc...
10    https://veganuary.com/recipes/categories/chine...
11    https://veganuary.com/recipes/categories/vegan...
12    https://veganuary.com/recipes/categories/irish...
13    https://veganuary.com/recipes/categories/healt...
14    https://veganuary.com/recipes/categories/briti...
15    https://veganuary.com/recipes/categories/greek...
16    https://veganuary.com/recipes/categories/portu...
17    https://veganuary.com/recipes/categories/t

In [5]:
categories["category"] = categories["category"].replace(r'https://veganuary.com/recipes/categories/', '', regex=True).replace(r'-vegan-recipes/', '', regex=True).replace(r'-recipes/', '', regex=True ).replace(r'vegan-', '', regex=True )


#websites['SERVER'] = websites['SERVER'].replace(r".*microsoft.*", "microsoft", regex=True).replace(r".*apache.*", "apache", regex=True).replace(r".*nginx.*", "nginx", regex=True)



In [6]:
categories

category_id                category  \
0             1                 italian   
1             2              quick-easy   
2             3           north-african   
3             4                japanese   
4             5                 seafood   
5             6                  korean   
6             7          south-american   
7             8                american   
8             9                  indian   
9            10                  french   
10           11                 chinese   
11           12                  easter   
12           13                   irish   
13           14                 healthy   
14           15                 british   
15           16                   greek   
16           17              portuguese   
17           18  turkish-middle-eastern   
18           19   thai-south-east-asian   
19           20                 one-pot   
20           21                 mexican   
21           22                  german   
22           23                 spanish   
23           24                 chicken   
24           25                 african   
25           26               caribbean   
26           27               christmas   

                                       category_links  
0   https://veganuary.com/recipes/categories/itali...  
1   https://veganuary.com/recipes/categories/quick...  
2   https://veganuary.com/recipes/categories/north...  
3   https://veganuary.com/recipes/categories/japan...  
4   https://veganuary.com/recipes/categories/vegan...  
5   https://veganuary.com/recipes/categories/korea...  
6   https://veganuary.com/recipes/categories/south...  
7   https://veganuary.com/recipes/categories/ameri...  
8   https://veganuary.com/recipes/categories/india...  
9   https://veganuary.com/recipes/categories/frenc...  
10  https://veganuary.com/recipes/categories/chine...  
11  https://veganuary.com/recipes/categories/vegan...  
12  https://veganuary.com/recipes/categories/irish...  
13  https://veganuary.com/recipes/categories/healt...  
14  https://veganuary.com/recipes/categories/briti...  
15  https://veganuary.com/recipes/categories/greek...  
16  https://veganuary.com/recipes/categories/portu...  
17  https://veganuary.com/recipes/categories/turki...  
18  https://veganuary.com/recipes/categories/thai-...  
19  https://veganuary.com/recipes/categories/one-p...  
20  https://veganuary.com/recipes/categories/mexic...  
21  https://veganuary.com/recipes/categories/germa...  
22  https://veganuary.com/recipes/categories/spani...  
23  https://veganuary.com/recipes/categories/vegan...  
24  https://veganuary.com/recipes/categories/afric...  
25  https://veganuary.com/recipes/categories/carib...  
26  https://veganuary.com/recipes/categories/vegan...

In [23]:
categories["category_id"] = categories.index + 1

In [27]:
categories = categories[['category_id', 'category', 'category_links']]



In [7]:
categories

category_id                category  \
0             1                 italian   
1             2              quick-easy   
2             3           north-african   
3             4                japanese   
4             5                 seafood   
5             6                  korean   
6             7          south-american   
7             8                american   
8             9                  indian   
9            10                  french   
10           11                 chinese   
11           12                  easter   
12           13                   irish   
13           14                 healthy   
14           15                 british   
15           16                   greek   
16           17              portuguese   
17           18  turkish-middle-eastern   
18           19   thai-south-east-asian   
19           20                 one-pot   
20           21                 mexican   
21           22                  german   
22           23                 spanish   
23           24                 chicken   
24           25                 african   
25           26               caribbean   
26           27               christmas   

                                       category_links  
0   https://veganuary.com/recipes/categories/itali...  
1   https://veganuary.com/recipes/categories/quick...  
2   https://veganuary.com/recipes/categories/north...  
3   https://veganuary.com/recipes/categories/japan...  
4   https://veganuary.com/recipes/categories/vegan...  
5   https://veganuary.com/recipes/categories/korea...  
6   https://veganuary.com/recipes/categories/south...  
7   https://veganuary.com/recipes/categories/ameri...  
8   https://veganuary.com/recipes/categories/india...  
9   https://veganuary.com/recipes/categories/frenc...  
10  https://veganuary.com/recipes/categories/chine...  
11  https://veganuary.com/recipes/categories/vegan...  
12  https://veganuary.com/recipes/categories/irish...  
13  https://veganuary.com/recipes/categories/healt...  
14  https://veganuary.com/recipes/categories/briti...  
15  https://veganuary.com/recipes/categories/greek...  
16  https://veganuary.com/recipes/categories/portu...  
17  https://veganuary.com/recipes/categories/turki...  
18  https://veganuary.com/recipes/categories/thai-...  
19  https://veganuary.com/recipes/categories/one-p...  
20  https://veganuary.com/recipes/categories/mexic...  
21  https://veganuary.com/recipes/categories/germa...  
22  https://veganuary.com/recipes/categories/spani...  
23  https://veganuary.com/recipes/categories/vegan...  
24  https://veganuary.com/recipes/categories/afric...  
25  https://veganuary.com/recipes/categories/carib...  
26  https://veganuary.com/recipes/categories/vegan...

In [29]:
categories.to_csv('veganuary_recipe_category_links.csv', index=False)


In [8]:
categories.shape

(27, 3)

In [9]:
list_cat_df = categories

In [10]:
list_cat = list_cat_df["category_links"]

In [33]:
list_cat 

0     https://veganuary.com/recipes/categories/itali...
1     https://veganuary.com/recipes/categories/quick...
2     https://veganuary.com/recipes/categories/north...
3     https://veganuary.com/recipes/categories/japan...
4     https://veganuary.com/recipes/categories/vegan...
5     https://veganuary.com/recipes/categories/korea...
6     https://veganuary.com/recipes/categories/south...
7     https://veganuary.com/recipes/categories/ameri...
8     https://veganuary.com/recipes/categories/india...
9     https://veganuary.com/recipes/categories/frenc...
10    https://veganuary.com/recipes/categories/chine...
11    https://veganuary.com/recipes/categories/vegan...
12    https://veganuary.com/recipes/categories/irish...
13    https://veganuary.com/recipes/categories/healt...
14    https://veganuary.com/recipes/categories/briti...
15    https://veganuary.com/recipes/categories/greek...
16    https://veganuary.com/recipes/categories/portu...
17    https://veganuary.com/recipes/categories/t

In [11]:
# getting recipes links  from categories

website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""
print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_links = []
recipes_category = []

driver = webdriver.Chrome()

for url in list_cat: 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_links.append(link.get_attribute('href'))
        recipes_category.append(url)

driver.quit()

recipe_links = list(zip(recipe_links, recipes_category ))

df = pd.DataFrame(recipe_links, columns=["recipes_category", "Category Links", ])

csv_file = 'veganuary_recipe_links2.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 10:15:45.600299
Crawl finished. CSV file saved as: veganuary_recipe_links2.csv
Crawl time: 0:02:55.375637


# getting recipes links from meal


In [2]:


breakfast1 = "https://veganuary.com/recipes/meals/breakfast/"
breakfast = [f"https://veganuary.com/recipes/meals/breakfast/page/{pag}" for pag in range(2,10)]
breakfast.append(breakfast1)

lunch1 = "https://veganuary.com/recipes/meals/lunch/page/"
lunch = [f"https://veganuary.com/recipes/meals/lunch/page/{pag}" for pag in range(2,22)]
lunch.append(lunch1)


diner1 = "https://veganuary.com/recipes/meals/dinner/"
diner = [f"https://veganuary.com/recipes/meals/dinner/page/{pag}" for pag in range(2,33)]
diner.append(diner1)


snacks_1 = "https://veganuary.com/recipes/meals/snacks/"
snacks = [f"https://veganuary.com/recipes/meals/snacks/page/{pag}" for pag in range(1,2)]
snacks.append(snacks_1)


baking_deserts1 = "https://veganuary.com/recipes/meals/sweet-treats/"
baking_deserts = [f"https://veganuary.com/recipes/meals/sweet-treats/page/{pag}" for pag in range(1,5)]
baking_deserts.append(baking_deserts1)


#list_meals = zip()

In [4]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""

print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_url = []
meal_category = []

driver = webdriver.Chrome()

for url in breakfast: 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_url.append(link.get_attribute('href'))
        meal_category.append("breakfast")

driver.quit()

recipe_links = list(zip( meal_category, recipe_url ))

df = pd.DataFrame(recipe_links, columns=["meal_category", "recipe_url" ])

csv_file = 'breakfast.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 11:05:11.165749
Crawl finished. CSV file saved as: breakfast.csv
Crawl time: 0:01:09.556712


In [5]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""

print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_url = []
meal_category = []

driver = webdriver.Chrome()

for url in lunch : 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_url.append(link.get_attribute('href'))
        meal_category.append("lunch ")

driver.quit()

recipe_links = list(zip( meal_category, recipe_url ))

df = pd.DataFrame(recipe_links, columns=["meal_category", "recipe_url" ])

csv_file = 'lunch.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 11:06:38.317136
Crawl finished. CSV file saved as: lunch.csv
Crawl time: 0:02:50.796461


In [6]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""

print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_url = []
meal_category = []

driver = webdriver.Chrome()

for url in diner: 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_url.append(link.get_attribute('href'))
        meal_category.append("diner")

driver.quit()

recipe_links = list(zip( meal_category, recipe_url ))

df = pd.DataFrame(recipe_links, columns=["meal_category", "recipe_url" ])

csv_file = 'diner.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 11:10:28.219977
Crawl finished. CSV file saved as: diner.csv
Crawl time: 0:04:18.298298


In [7]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""

print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_url = []
meal_category = []

driver = webdriver.Chrome()

for url in snacks: 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_url.append(link.get_attribute('href'))
        meal_category.append("snacks")

driver.quit()

recipe_links = list(zip( meal_category, recipe_url ))

df = pd.DataFrame(recipe_links, columns=["meal_category", "recipe_url" ])

csv_file = 'snacks.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 11:16:07.278262
Crawl finished. CSV file saved as: snacks.csv
Crawl time: 0:00:21.690284


In [8]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
"""

print(website)

start_time = datetime.now()
print('Crawl starting time:', start_time.time())

recipe_url = []
meal_category = []

driver = webdriver.Chrome()

for url in baking_deserts: 
    driver.get(url)
    sleep(5)
    all_links = driver.find_elements(By.XPATH, '//a[contains(@href, "/recipes/") and not(contains(@href, "/categories/"))]')
    for link in all_links:
        recipe_url.append(link.get_attribute('href'))
        meal_category.append("baking_deserts")

driver.quit()

recipe_links = list(zip( meal_category, recipe_url ))

df = pd.DataFrame(recipe_links, columns=["meal_category", "recipe_url" ])

csv_file = 'baking_deserts.csv'
df.to_csv(csv_file, index=False)


print('Crawl finished. CSV file saved as:', csv_file)
print('Crawl time:', datetime.now() - start_time)



########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################

Crawl starting time: 11:17:05.985505
Crawl finished. CSV file saved as: baking_deserts.csv
Crawl time: 0:00:40.125643


In [22]:
df_breakfast = pd.read_csv("breakfast.csv")
df_lunch = pd.read_csv("lunch.csv")
df_diner = pd.read_csv("diner.csv")
df_snacks = pd.read_csv("snacks.csv")
df_deserts =  pd.read_csv("baking_deserts.csv")

In [78]:
df_breakfast["meal_id"] = 1
df_lunch["meal_id"] = 2
df_diner ["meal_id"] = 3
df_snacks["meal_id"] = 4
df_deserts["meal_id"] = 5

In [134]:
df_breakfast

meal_category                                         recipe_url  meal_id
0       breakfast                     https://veganuary.com/recipes/        1
1       breakfast                     https://veganuary.com/recipes/        1
2       breakfast     https://veganuary.com/recipes/meals/breakfast/        1
3       breakfast         https://veganuary.com/recipes/meals/lunch/        1
4       breakfast        https://veganuary.com/recipes/meals/dinner/        1
..            ...                                                ...      ...
326     breakfast  https://veganuary.com/recipes/fluffy-peanut-bu...        1
327     breakfast  https://veganuary.com/recipes/meals/breakfast/...        1
328     breakfast  https://veganuary.com/recipes/meals/breakfast/...        1
329     breakfast  https://veganuary.com/recipes/meals/breakfast/...        1
330     breakfast  https://veganuary.com/recipes/meals/breakfast/...        1

[331 rows x 3 columns]

In [80]:
df_meals = pd.concat([df_breakfast, df_lunch, df_diner, df_snacks, df_deserts], axis=0)

In [81]:
df_meals.shape

(2646, 3)

In [82]:
df_meals["recipe_url"].duplicated().sum()

1719

In [84]:
df_meals.drop_duplicates(subset="recipe_url", keep="first", inplace=True)


In [85]:
df_meals["recipe_url"].duplicated().sum()

0

In [135]:
df_meals.shape

(921, 6)

In [138]:
df_meals = df_meals[~df_meals['url'].str.contains("/page/")]


In [17]:
mapping_meal = {"breakfast": 1, "lunch": 2, "diner": 3, "snacks": 4, "baking_deserts": 5}


meals = pd.DataFrame(mapping_meal.items(), columns=['meal', 'meal_id'])


In [18]:
meals = meals[['meal_id', 'meal']]

In [100]:
df_meals["website_id"] = 2

In [20]:
meals.to_csv("meal_ids.csv")

In [86]:
df_meals.head(10)

meal_category                                         recipe_url  meal_id
0      breakfast                     https://veganuary.com/recipes/        1
2      breakfast     https://veganuary.com/recipes/meals/breakfast/        1
3      breakfast         https://veganuary.com/recipes/meals/lunch/        1
4      breakfast        https://veganuary.com/recipes/meals/dinner/        1
5      breakfast        https://veganuary.com/recipes/meals/snacks/        1
6      breakfast  https://veganuary.com/recipes/meals/sweet-treats/        1
14     breakfast  https://veganuary.com/recipes/chocolate-protei...        1
15     breakfast  https://veganuary.com/recipes/choc-peanut-butt...        1
16     breakfast  https://veganuary.com/recipes/vegan-tofu-bened...        1
17     breakfast      https://veganuary.com/recipes/overnight-oats/        1

In [139]:
values_to_drop =  ["https://veganuary.com/recipes/", "https://veganuary.com/recipes/meals/breakfast/", "https://veganuary.com/recipes/meals/lunch/", "https://veganuary.com/recipes/meals/dinner/", "https://veganuary.com/recipes/meals/snacks/", "https://veganuary.com/recipes/meals/sweet-treats/"]

In [141]:
df_meals = df_meals[~df_meals['url'].isin(values_to_drop)]




In [142]:
df_meals["url"].duplicated().sum()

0

In [143]:
df_meals.shape

(849, 6)

In [145]:
df_meals.drop(columns=["recipe_id"])

meal_id   meal_category  \
14         1       breakfast   
15         1       breakfast   
16         1       breakfast   
17         1       breakfast   
18         1       breakfast   
..       ...             ...   
139        5  baking_deserts   
140        5  baking_deserts   
142        5  baking_deserts   
143        5  baking_deserts   
144        5  baking_deserts   

                                                   url  website_id  diet_id  
14   https://veganuary.com/recipes/chocolate-protei...           2        1  
15   https://veganuary.com/recipes/choc-peanut-butt...           2        1  
16   https://veganuary.com/recipes/vegan-tofu-bened...           2        1  
17       https://veganuary.com/recipes/overnight-oats/           2        1  
18    https://veganuary.com/recipes/aloo-masala-toast/           2        1  
..                                                 ...         ...      ...  
139  https://veganuary.com/recipes/vegan-chocolate-...           2        1  
140  https://veganuary.com/recipes/perfectly-vanill...           2        1  
142  https://veganuary.com/recipes/vegan-hot-cross-...           2        1  
143  https://veganuary.com/recipes/vegan-pavlova-wi...           2        1  
144  https://veganuary.com/recipes/chocolate-orange...           2        1  

[849 rows x 5 columns]

In [101]:
df_meals.head()

meal_id meal_category                                                url  \
14        1     breakfast  https://veganuary.com/recipes/chocolate-protei...   
15        1     breakfast  https://veganuary.com/recipes/choc-peanut-butt...   
16        1     breakfast  https://veganuary.com/recipes/vegan-tofu-bened...   
17        1     breakfast      https://veganuary.com/recipes/overnight-oats/   
18        1     breakfast   https://veganuary.com/recipes/aloo-masala-toast/   

    website_id  
14           2  
15           2  
16           2  
17           2  
18           2

In [105]:
df_meals = df_meals[['meal_id', 'meal_category', "url", "website_id"]]



In [160]:
df_meals

meal_id   meal_category  \
0          1       breakfast   
1          1       breakfast   
2          1       breakfast   
3          1       breakfast   
4          1       breakfast   
..       ...             ...   
844        5  baking_deserts   
845        5  baking_deserts   
846        5  baking_deserts   
847        5  baking_deserts   
848        5  baking_deserts   

                                                   url  website_id  diet_id  
0    https://veganuary.com/recipes/chocolate-protei...           2        1  
1    https://veganuary.com/recipes/choc-peanut-butt...           2        1  
2    https://veganuary.com/recipes/vegan-tofu-bened...           2        1  
3        https://veganuary.com/recipes/overnight-oats/           2        1  
4     https://veganuary.com/recipes/aloo-masala-toast/           2        1  
..                                                 ...         ...      ...  
844  https://veganuary.com/recipes/vegan-chocolate-...           2        1  
845  https://veganuary.com/recipes/perfectly-vanill...           2        1  
846  https://veganuary.com/recipes/vegan-hot-cross-...           2        1  
847  https://veganuary.com/recipes/vegan-pavlova-wi...           2        1  
848  https://veganuary.com/recipes/chocolate-orange...           2        1  

[849 rows x 5 columns]

In [161]:
df_meals["recipe_id"] = "VN" + (df_meals.index + 1).astype(str)

In [162]:
df_meals

meal_id   meal_category  \
0          1       breakfast   
1          1       breakfast   
2          1       breakfast   
3          1       breakfast   
4          1       breakfast   
..       ...             ...   
844        5  baking_deserts   
845        5  baking_deserts   
846        5  baking_deserts   
847        5  baking_deserts   
848        5  baking_deserts   

                                                   url  website_id  diet_id  \
0    https://veganuary.com/recipes/chocolate-protei...           2        1   
1    https://veganuary.com/recipes/choc-peanut-butt...           2        1   
2    https://veganuary.com/recipes/vegan-tofu-bened...           2        1   
3        https://veganuary.com/recipes/overnight-oats/           2        1   
4     https://veganuary.com/recipes/aloo-masala-toast/           2        1   
..                                                 ...         ...      ...   
844  https://veganuary.com/recipes/vegan-chocolate-...           2        1   
845  https://veganuary.com/recipes/perfectly-vanill...           2        1   
846  https://veganuary.com/recipes/vegan-hot-cross-...           2        1   
847  https://veganuary.com/recipes/vegan-pavlova-wi...           2        1   
848  https://veganuary.com/recipes/chocolate-orange...           2        1   

    recipe_id  
0         VN1  
1         VN2  
2         VN3  
3         VN4  
4         VN5  
..        ...  
844     VN845  
845     VN846  
846     VN847  
847     VN848  
848     VN849  

[849 rows x 6 columns]

In [163]:
df_meals = df_meals[["recipe_id", 'meal_id', 'meal_category', "url", "website_id"]]

In [164]:
df_meals.head()

recipe_id  meal_id meal_category  \
0       VN1        1     breakfast   
1       VN2        1     breakfast   
2       VN3        1     breakfast   
3       VN4        1     breakfast   
4       VN5        1     breakfast   

                                                 url  website_id  
0  https://veganuary.com/recipes/chocolate-protei...           2  
1  https://veganuary.com/recipes/choc-peanut-butt...           2  
2  https://veganuary.com/recipes/vegan-tofu-bened...           2  
3      https://veganuary.com/recipes/overnight-oats/           2  
4   https://veganuary.com/recipes/aloo-masala-toast/           2

In [133]:
df_meals["url"]

14     https://veganuary.com/recipes/chocolate-protei...
15     https://veganuary.com/recipes/choc-peanut-butt...
16     https://veganuary.com/recipes/vegan-tofu-bened...
17         https://veganuary.com/recipes/overnight-oats/
18      https://veganuary.com/recipes/aloo-masala-toast/
                             ...                        
140    https://veganuary.com/recipes/perfectly-vanill...
142    https://veganuary.com/recipes/vegan-hot-cross-...
143    https://veganuary.com/recipes/vegan-pavlova-wi...
144    https://veganuary.com/recipes/chocolate-orange...
150    https://veganuary.com/recipes/meals/sweet-trea...
Name: url, Length: 921, dtype: object

In [115]:
df_meals

recipe_id  meal_id   meal_category  \
14       VN15        1       breakfast   
15       VN16        1       breakfast   
16       VN17        1       breakfast   
17       VN18        1       breakfast   
18       VN19        1       breakfast   
..        ...      ...             ...   
140     VN141        5  baking_deserts   
142     VN143        5  baking_deserts   
143     VN144        5  baking_deserts   
144     VN145        5  baking_deserts   
150     VN151        5  baking_deserts   

                                                   url  website_id  diet_id  
14   https://veganuary.com/recipes/chocolate-protei...           2        1  
15   https://veganuary.com/recipes/choc-peanut-butt...           2        1  
16   https://veganuary.com/recipes/vegan-tofu-bened...           2        1  
17       https://veganuary.com/recipes/overnight-oats/           2        1  
18    https://veganuary.com/recipes/aloo-masala-toast/           2        1  
..                                                 ...         ...      ...  
140  https://veganuary.com/recipes/perfectly-vanill...           2        1  
142  https://veganuary.com/recipes/vegan-hot-cross-...           2        1  
143  https://veganuary.com/recipes/vegan-pavlova-wi...           2        1  
144  https://veganuary.com/recipes/chocolate-orange...           2        1  
150  https://veganuary.com/recipes/meals/sweet-trea...           2        1  

[921 rows x 6 columns]

In [114]:
df_meals["diet_id"] = 1

In [165]:
#df_meals = df_meals.rename(columns={"recipe_url": "url"})
df_meals.to_csv("veganuary_meal_recipes.csv")

The dataframe has the list of 921 URLS with recipes, bigger than the other categorization. THis one will be used to scrap more recipes

In [166]:
df_meals.shape

(849, 5)

In [167]:
reci_li = list(df_meals["url"])

In [168]:
len(reci_li)

849

In [169]:
reci_li

['https://veganuary.com/recipes/chocolate-protein-pancakes/',
 'https://veganuary.com/recipes/choc-peanut-butter-porridge/',
 'https://veganuary.com/recipes/vegan-tofu-benedict/',
 'https://veganuary.com/recipes/overnight-oats/',
 'https://veganuary.com/recipes/aloo-masala-toast/',
 'https://veganuary.com/recipes/tofu-bhurji-indian-scrambled-tofu/',
 'https://veganuary.com/recipes/berry-blast-smoothie/',
 'https://veganuary.com/recipes/apple-and-pear-overnight-oats/',
 'https://veganuary.com/recipes/buckwheat-banana-pancakes/',
 'https://veganuary.com/recipes/vegan-horlicks-porridge/',
 'https://veganuary.com/recipes/alpro-vegan-overnight-oats/',
 'https://veganuary.com/recipes/sausage-spinach-avocado-muffin/',
 'https://veganuary.com/recipes/bright-barley-double-chocolate-banana-french-toast/',
 'https://veganuary.com/recipes/matcha-pancakes/',
 'https://veganuary.com/recipes/apple-cinnamon-spiced-protein-oats/',
 'https://veganuary.com/recipes/peaches-cream-baked-oats/',
 'https://ve

In [ ]:


website = """
########################################################
#              WEBSITE: https://veganuary.com/recipes/         #
########################################################
#"""
print(website)

urls = []
recipes = []
ingredients_list = []
instructions = []
diet = []
serves = []
time_prep = []
time_cook = []

tart_time = datetime.now()
print('Crawl starting time:', start_time.time())



reci_list = reci_li

driver = webdriver.Chrome()



urls = []
recipes = []
ingredients_list = []
instructions = []
diet = []
serves = []
time_prep = []
time_cook = []

for url in reci_list:
    try:
        print('Working on', url)
        driver.get(url)
        sleep(randint(3,6))  
        
    
        title_element = WebDriverWait(driver, 10).until(
            EC.visibility_of_element_located((By.XPATH, '//h1[@class="hero__title"]'))
        )
        title_text = title_element.text
        recipes.append(title_text)
        
    
        prep_time_element = driver.find_element(By.XPATH, '//div[@class="tooltip-single"][1]/div[@class="tooltip-text-recipe"]')
        prep_time = prep_time_element.text.split(': ')[1]
        time_prep.append(prep_time)

        cook_time_element = driver.find_element(By.XPATH, '//div[@class="tooltip-single"][2]/div[@class="tooltip-text-recipe"]')
        cook_time_e = cook_time_element.text.split(': ')[1]
        time_cook.append(cook_time_e)

    
        servings_element = driver.find_element(By.XPATH, '//div[@class="tooltip-single"][3]/div[@class="tooltip-text-recipe"]')
        servings = servings_element.text.split(': ')[1]  
        serves.append(servings)
    
        ingred_element1 = driver.find_element(By.XPATH, '//div[@class="recipe__ingredients"]')
        ingred_element = ingred_element1.text 
        ingredients_list.append(ingred_element)

        prep_element1 = driver.find_element(By.XPATH, '//div[@class="col-lg-8"]')
        prep_element = prep_element1.text 
        instructions.append(prep_element)


        diet.append('1')
        urls.append(url)

    except Exception as e:
        print(f'Recipe {url} does not exist or could not be processed. Error:', e)


data2 = list(zip(urls, recipes, ingredients_list, instructions, diet, serves, time_prep, time_cook))
df2 = pd.DataFrame(data2, columns=["url", "recipes", "list_ingredients", "instructions", "diet", "serves", "time_prep","time_cook" ])


driver.quit()


csv_file = 'veganuary_recipe_trytry.csv'
df2.to_csv(csv_file, index=False)

#print('Crawl finished. CSV file saved as:', csv_file)
#print('Crawl time:', datetime.now() - start_time)


In [1]:
df2.shape

NameError: name 'df2' is not defined